### ISSUE TITLE
Change return `True`/`False` pattern to `try/catch` pattern

### ISSUE THREAD

**Issue by NIC619** [libp2p/py-libp2p#194]
Change pattern like this(L59): https://github.com/libp2p/py-libp2p/blob/f0046fa3e0952b492837a698b1988e05c9821f47/libp2p/host/basic_host.py#L54-L59

to try/ catch pattern

**Comment by ralexstokes** [comment]
what is the rationale here?

**Comment by NIC619** [comment]
Just align with the error handling pattern of python IMO?

**Comment by ralexstokes** [comment]
gotcha -- yeah pythonistas do like their exceptions :)

my own take on it is that exceptions don't compose very well which leads to several symptoms in code like non-local effects/interactions that make code harder to reason about. i also cant take library code that may throw an exception and put it into other arbitrary code w/o also considering how the exception will affect my new code -- again, higher mental load/context. if i only have to consider (and handle!) a single boolean right at the call site, it tends to keep things local, even possibly at the expense of more verbose code (altho i think in this case the verbosity is better than the non-local effects)

while we are here, we could also have a conversation about the implicit mutability of this API but that is not what this issue is about so i'll continue on :)

i'm not super against using exceptions where appropriate but i'd think we need to consider on a case-by-case basis, rather than a blanket policy

**Comment by NIC619** [comment]
Agree on the mental load and verbosity/non-local effects trade-off.

These are the callees that returns true/false:
- `Network.set_stream_handler()`
- `Network.listen()`
- `Network.notify()`
- `NetStream.set_protocol()`
- `MplexStream.close()`
- `MplexStream.reset()`
- `MplexStream.set_*_deadline()`

IMO only `Network.listen()`, `MplexStream.close()` and `MplexStream.reset()` need to have a return value(or throwing exceptions).
- For `Network.listen()`
https://github.com/libp2p/py-libp2p/blob/21e013e7537d4cad66a5d73c21e05d54a40ba33e/libp2p/network/network_interface.py#L61-L65
    - Since there are possibly more than one multiaddr to listen on and only one of them we need to listen on successfully, I'm +1 on return true/false here.
- For `MplexStream.close()` and `MplexStream.reset()`
    - Not sure about this. What do we do when closing/resetting a stream failed? It seems to me more verbosity is needed in this case but probably have to figure out what could go wrong when closing/resetting a stream first.

**Comment by mhchia** [comment]
Is this solved yet?

**Comment by ralexstokes** [comment]
i dont think so -- i'd support a move to exceptions where it makes sense but its not a top priority for me right now

**Comment by BlacAc3** [comment]
Hello, I believe this issue is resolved and has not been closed?
